src/Domain/ElementDistribution.hpp

```diff
@@ -0,0 +1,71 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <array>
+#include <cstddef>
+#include <utility>
+#include <vector>
+
+#include "Domain/Structure/ElementId.hpp"
+
+namespace domain {
+
+/*!
+ * \brief Distribution strategy based on grouping elements within blocks
+ * together, and using a 'Z-curve' to determine distribution within blocks.
+ *
+ * \details This class suggests starting placements of elements on processors by
+ * first greedily assigning portions of blocks to each processor, then dividing
+ * up the blocks among processors that share it according to a 'Z-curve'.
+ * Z-curves are a simple and easily-computed space-filling curve that (unlike
+ * hilbert curves) permit diagonal traversal.
+ *
+ * For example, here is a sketch of a 4x2 2D block, with each element labelled
+ * according to the order on the Z-curve:
+ * ```
+ *          x-->
+ *          0   1   2   3
+ *        ----------------
+ *  y  0 |  0   2   4   6
+ *  |    |  | / | / | / |
+ *  v  1 |  1   3   5   7
+ * ```
+ * (forming a zig-zag path, that under some rotation/reflection has a 'Z'
+ * shape).
+ *
+ * Then, the first processor for a block takes as many elements have been
```

Comment:
The two large clusters statement is, I believe, general. I haven't constructed a theorem or anything, but it is part of the test and I haven't thought of any counterexamples.
Here's a rough argument:
So, consider if we have some starting index somewhere in the middle of the Z-curve in binary representation. It goes from that point to some other index in binary representation. It will have a most significant bit change, and all of the less significant other bits will carry through an appropriate increment sequence, so form a dense set from the start sequence to the most significant bit change, and from that most significant bit change to the end of the sequence. The most significant bit change represents a 'jump' in spatial location along one of the axes, but the first and second halves (before and after the most significant bit change) I'm pretty sure are guaranteed to be orthogonally connected chunks -- tracing the curve forward and back from that 'largest jump' results in only connected parts.
(there's probably a similarly decent geometrical argument but I found it easier to think in bits)

I'll try to clarify the docs without making the discussion too inscrutable :P